fedimint-api/src/core.rs

```diff
@@ -211,45 +243,72 @@ macro_rules! newtype_impl_display_passthrough {
+    type Output: PluginOutput;
+    type OutputOutcome: PluginOutputOutcome;
+    type ConsensusItem: PluginConsensusItem;
+
     /// Decode `Input` compatible with this module, after the module key prefix was already decoded
-    fn decode_input(r: &mut dyn io::Read) -> Result<Input, DecodeError>;
+    fn decode_input(&self, r: &mut dyn io::Read) -> Result<Self::Input, DecodeError>;
 
     /// Decode `Output` compatible with this module, after the module key prefix was already decoded
-    fn decode_output(r: &mut dyn io::Read) -> Result<Output, DecodeError>;
+    fn decode_output(&self, r: &mut dyn io::Read) -> Result<Self::Output, DecodeError>;
 
     /// Decode `OutputOutcome` compatible with this module, after the module key prefix was already decoded
-    fn decode_output_outcome(r: &mut dyn io::Read) -> Result<OutputOutcome, DecodeError>;
+    fn decode_output_outcome(
+        &self,
+        r: &mut dyn io::Read,
+    ) -> Result<Self::OutputOutcome, DecodeError>;
 
     /// Decode `ConsensusItem` compatible with this module, after the module key prefix was already decoded
-    fn decode_consensus_item(r: &mut dyn io::Read) -> Result<ConsensusItem, DecodeError>;
+    fn decode_consensus_item(
+        &self,
+        r: &mut dyn io::Read,
+    ) -> Result<Self::ConsensusItem, DecodeError>;
 }
 
 pub trait ModuleDecode: Debug {
     /// Decode `Input` compatible with this module, after the module key prefix was already decoded
-    fn decode_input(&self, r: &mut dyn io::Read) -> Result<Input, DecodeError>;
+    fn decode_input(
+        &self,
+        r: &mut dyn io::Read,
+        instance_id: ModuleInstanceId,
+    ) -> Result<Input, DecodeError>;
 
     /// Decode `Output` compatible with this module, after the module key prefix was already decoded
-    fn decode_output(&self, r: &mut dyn io::Read) -> Result<Output, DecodeError>;
+    fn decode_output(
+        &self,
+        r: &mut dyn io::Read,
+        instance_id: ModuleInstanceId,
+    ) -> Result<Output, DecodeError>;
 
     /// Decode `OutputOutcome` compatible with this module, after the module key prefix was already decoded
-    fn decode_output_outcome(&self, r: &mut dyn io::Read) -> Result<OutputOutcome, DecodeError>;
+    fn decode_output_outcome(
+        &self,
+        r: &mut dyn io::Read,
+        instance_id: ModuleInstanceId,
+    ) -> Result<OutputOutcome, DecodeError>;
 
     /// Decode `ConsensusItem` compatible with this module, after the module key prefix was already decoded
-    fn decode_consensus_item(&self, r: &mut dyn io::Read) -> Result<ConsensusItem, DecodeError>;
+    fn decode_consensus_item(
+        &self,
+        r: &mut dyn io::Read,
+        instance_id: ModuleInstanceId,
+    ) -> Result<ConsensusItem, DecodeError>;
 }
 
 // TODO: use macro again
 /// Decoder for module associated types
 #[derive(Clone)]
-pub struct Decoder(&'static (dyn ModuleDecode + Send + Sync));
+pub struct Decoder(Arc<dyn ModuleDecode + Send + Sync>);
```

Comment:
Ok, you won :stuck_out_tongue_closed_eyes:

Reply:
:D